LykkeApi2/Controllers/AffiliateController.cs

```diff
@@ -0,0 +1,128 @@
+    {
+        private readonly IAffiliateClient _affiliateClient;
+        private readonly IRequestContext _requestContext;
+        private readonly IRateCalculatorClient _rateCalculatorClient;
+        private readonly ILog _log;
+
+        public AffiliateController(IAffiliateClient affiliateClient, IRequestContext requestContext, IRateCalculatorClient rateCalculatorClient, ILog log)
+        {
+            _affiliateClient = affiliateClient;
+            _requestContext = requestContext;
+            _rateCalculatorClient = rateCalculatorClient;
+            _log = log;
+        }
+
+        [HttpGet]
+        [Route("link")]
+        [ProducesResponseType(typeof(AffiliateLinkResponse), (int)HttpStatusCode.OK)]
+        [ProducesResponseType((int)HttpStatusCode.NotFound)]
+        public async Task<IActionResult> GetLink()
+        {
+            var links = await _affiliateClient.GetLinks(_requestContext.ClientId);
+            var link = links.FirstOrDefault();
+
+            if (link != null)
+            {
+                return Ok(new AffiliateLinkResponse
+                {
+                    Url = link.Url,
+                    RedirectUrl = link.RedirectUrl
+                });
+            }
+            else
+            {
+                return NotFound();
+            }
+        }
+
+        [HttpPost]
+        [Route("create")]
+        [ProducesResponseType(typeof(AffiliateLinkResponse), (int)HttpStatusCode.OK)]
+        [ProducesResponseType((int)HttpStatusCode.BadRequest)]
+        public async Task<IActionResult> CreateLink()
+        {
+            LinkModel link;
+            try
+            {
+                const string redirectUrl = "https://lykke.com";
+                link = await _affiliateClient.RegisterLink(_requestContext.ClientId, redirectUrl);
+            }
+            catch (HttpOperationException e)
+            {
+                await _log.WriteErrorAsync(nameof(AffiliateController), nameof(GetStats), e);
+                return BadRequest(new { message = e.Response.Content });
+            }
+
+            await _log.WriteInfoAsync(nameof(AffiliateController), nameof(GetStats), 
+                $"ClientId: {_requestContext.ClientId}. Url: {link.Url}. RedirectUrl: {link.RedirectUrl}");
+
+            return Ok(new AffiliateLinkResponse
+            {
+                Url = link.Url,
+                RedirectUrl = link.RedirectUrl
+            });
+        }
+
+
+        [HttpGet]
+        [Route("stats")]
+        [ProducesResponseType(typeof(AffiliateStatisticsResponse), (int)HttpStatusCode.OK)]
+        public async Task<IActionResult> GetStats()
+        {
+            var referrals = await _affiliateClient.GetReferrals(_requestContext.ClientId);
+
+            var stats = (await _affiliateClient.GetStats(_requestContext.ClientId)).ToList();
+
+            var bonuses = await GetAmountInBtc(stats.ToDictionary(x => x.AssetId, x => x.BonusVolume));
```

Comment:
Maybe is it better to calculate bonuses in the affiliate service?

Reply:
no, add one more dependency to affiliate service is unnecessary, plus maybe later we won't calculate it and will show amount for each asset to user